lib/doorkeeper/models/access_token_mixin.rb

```diff
@@ -120,44 +106,113 @@ def find_access_token_in_batches(relation, **args, &block)
+        token = filter_by_application_scopes(token, application, scopes)
 
-        find_access_token_in_batches(relation, batch_size: batch_size) do |batch|
-          tokens = batch.select do |token|
-            scopes_match?(token.scopes, scopes, application&.scopes)
-          end
+        token.order(created_at: :desc).first
+      end
+
+      # Determines if a list of scopes, when compared to an applications
+      # scopes, is invalid.
+      #
+      # There is a special case where if the application has no scopes,
+      # that we accept it as valid.
+      #
+      # @param scopes [Doorkeeper::Oauth::Scopes]
+      #   set of scopes
+      # @param application [Doorkeeper::Application]
+      #   Application instance
+      #
+      # @return [Boolean] true if there is an invalid scope, false if
+      #   there are no invalid scopes (or if the application has no
+      #   scopes)
+      #
+      def includes_invalid_scope(scopes, application)
+        return false if application.scopes.all.count == 0
 
-          matching_tokens.concat(tokens)
+        scopes.each do |requested_scope|
+          return true unless application.scopes.exists? requested_scope
         end
 
-        matching_tokens.max_by(&:created_at)
+        false
       end
 
-      # Checks whether the token scopes match the scopes from the parameters
+      # Filters by acceptable application scopes.
       #
-      # @param token_scopes [#to_s]
-      #   set of scopes (any object that responds to `#to_s`)
-      # @param param_scopes [Doorkeeper::OAuth::Scopes]
-      #   scopes from params
-      # @param app_scopes [Doorkeeper::OAuth::Scopes]
-      #   Application scopes
-      #
-      # @return [Boolean] true if the param scopes match the token scopes,
-      #   and all the param scopes are defined in the application (or in the
-      #   server configuration if the application doesn't define any scopes),
-      #   and false in other cases
-      #
-      def scopes_match?(token_scopes, param_scopes, app_scopes)
-        return true if token_scopes.empty? && param_scopes.empty?
-
-        (token_scopes.sort == param_scopes.sort) &&
-          Doorkeeper::OAuth::Helpers::ScopeChecker.valid?(
-            scope_str: param_scopes.to_s,
-            server_scopes: Doorkeeper.config.scopes,
-            app_scopes: app_scopes,
-          )
+      # @param relation [ActiveRecord::Relation]
+      #   Access tokens relation
+      # @param application [Doorkeeper::Application]
+      #   Application instance
+      # @param scopes [String, Doorkeeper::OAuth::Scopes]
+      #   set of scopes
+      #
+      # @return [ActiveRecord::Relation] filtered relation
+      #   query
+      #
+      def filter_by_application_scopes(relation, application, scopes)
+        application.scopes.each do |scope_val|
+          next unless scopes.exists? scope_val
+
+          relation = relation.where([
```

Comment:
I don't sure actually that this implementation matches previous one with `Doorkeeper::OAuth::Helpers::ScopeChecker
` :thinking: Even with passing specs it looks scary.. I have to to a deep dive into the changes to understand if everything is OK

Reply:
Ok, I can see the differences related to the grant types, is that what you are worried about here? If so, I'll see about getting this updated.

I am worried about the fact that the existing tests enforce that if a client has no scopes, that we allow any scopes through, is that on purpose?